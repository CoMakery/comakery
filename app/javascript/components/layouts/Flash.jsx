import React from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import Icon from './../styleguide/Icon'
import { FLASH_ADD_MESSAGE } from '../../src/javascripts/eventTypes'
import PubSub from 'pubsub-js'


export default class Flash extends React.Component {
  constructor(props) {
    super(props)

    this.closeMessage = this.closeMessage.bind(this)
    this.addMessage = this.addMessage.bind(this)

    PubSub.subscribe(FLASH_ADD_MESSAGE, this.addMessage);

    this.state = {
      messages: this.props.messages
    }
  }

  componentWillReceiveProps(props) {
    this.setState({
      messages: props.messages
    })
  }

  addMessage(_, message) {
    let messages = this.state.messages;
    messages.push(message);

    this.setState({
      messages: messages
    })
  }

  closeMessage(i) {
    let messages = this.state.messages
    messages.splice(i, 1)

    this.setState({
      messages: messages
    })
  }

  render() {
    const { className } = this.props

    return (
      <React.Fragment>
        <div className="flash-message-container">
          {this.state.messages.map((message, i) =>
            <div
              key={i}
              className={classNames(
                'flash-message',
                `flash-message--${message.severity}`,
                className
              )}
            >
              <Icon
                className="flash-message__icon"
                name={`flash/${message.severity}.svg`}
              />

              <div
                className="flash-message__text"
                dangerouslySetInnerHTML={{ __html: message.text }}
              />

              <Icon
                className="flash-message__icon flash-message__icon--close"
                name="flash/close.svg"
                onClick={() => { this.closeMessage(i) }}
              />
            </div>
          )}
        </div>
      </React.Fragment>
    )
  }
}

Flash.propTypes = {
  className: PropTypes.string,
  messages : PropTypes.array
}
Flash.defaultProps = {
  className: '',
  messages : [
    {
      'severity': 'warning',
      'text'    : 'warning text'
    }
  ]
}
