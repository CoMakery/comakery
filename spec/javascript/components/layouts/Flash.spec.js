import React from 'react'
import { shallow } from 'enzyme'
import Flash from 'components/layouts/Flash'

describe('Flash', () => {
  it('renders correctly without props', () => {
    const wrapper = shallow(<Flash />)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.exists('.flash-message')).toBe(true)
    expect(wrapper.exists('.flash-message__icon')).toBe(true)
    expect(wrapper.exists('.flash-message__icon--close')).toBe(true)
  })

  it('renders correctly with notice flashType prop', () => {
    const wrapper = shallow(<Flash flashType="notice" />)

    expect(wrapper.exists('.flash-message--notice')).toBe(true)
  })

  it('renders correctly with error flashType prop', () => {
    const wrapper = shallow(<Flash flashType="error" />)

    expect(wrapper.exists('.flash-message--error')).toBe(true)
  })

  it('renders correctly with message prop', () => {
    const wrapper = shallow(<Flash flashType="notice" message="Test Message" />)

    expect(wrapper.exists('.flash-message--notice')).toBe(true)
    expect(wrapper.find('.flash-message').text()).toBe('<Icon />Test Message<Icon />')
  })
})
