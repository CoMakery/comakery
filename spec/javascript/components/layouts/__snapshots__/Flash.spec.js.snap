// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flash renders correctly without props 1`] = `
<Flash
  className=""
  messages={
    Array [
      Object {
        "severity": "warning",
        "text": "warning text",
      },
    ]
  }
>
  <div
    className="flash-message-container"
  >
    <div
      className="flash-message flash-message--warning"
      key="0"
    >
      <Icon
        className="flash-message__icon"
        name="flash/warning.svg"
      >
        <img
          className="icon-legacy icon__flash/warning-svg flash-message__icon"
          src="test-file-stub"
        />
      </Icon>
      <div
        className="flash-message__text"
        dangerouslySetInnerHTML={
          Object {
            "__html": "warning text",
          }
        }
      />
      <Icon
        className="flash-message__icon flash-message__icon--close"
        name="flash/close.svg"
        onClick={[Function]}
      >
        <img
          className="icon-legacy icon__flash/close-svg flash-message__icon flash-message__icon--close"
          onClick={[Function]}
          src="test-file-stub"
        />
      </Icon>
    </div>
  </div>
</Flash>
`;
